router/core/src/main/scala/io/buoyant/router/RoutingFactory.scala

```diff
@@ -16,6 +16,42 @@ object RoutingFactory {
    */
   type Identifier[Req] = Req => Future[RequestIdentification[Req]]
 
+  object Identifier {
+
+    def compose[Req](hd: Identifier[Req], tl: Identifier[Req]*): Identifier[Req] =
+      compose(hd +: tl)
+
+    /**
+     * Apply each identifier in order until the request is identified.
+     *
+     * Identifier list must not be empty.
+     */
+    def compose[Req](all: Seq[Identifier[Req]]): Identifier[Req] = all match {
+      case Nil => throw new IllegalArgumentException("empty identifier list")
+      case Seq(identifier) => identifier
+      case Seq(hd, tl@_*) => (req: Req) => fold(req, hd, tl)
+    }
+
+    /**
+     * Apply each identifier to the request in order until the request
+     * is identified or identifiers are exhausted.
+     */
+    private[this] def fold[Req](
```

Comment:
not with future.flatMap -- it's still "safe" in that this won't blow the stack, but tailrec and futures are not friends.